src/Entity/Type/RefundEnumType.php

```diff
@@ -30,31 +31,37 @@ public function getSQLDeclaration(array $fieldDeclaration, AbstractPlatform $pla
         return 'VARCHAR(256)';
     }
 
-    public function convertToPHPValue($value, AbstractPlatform $platform): RefundType
+    public function convertToPHPValue($value, AbstractPlatform $platform): RefundTypeInterface
     {
-        if (!RefundType::isValid($value)) {
+        if ($value instanceof RefundType && !$value::isValid($value)) {
             throw new \InvalidArgumentException(sprintf(
                 'The value "%s" is not valid for the enum "%s". Expected one of ["%s"]',
                 $value,
-                RefundType::class,
-                implode('", "', RefundType::keys())
-            ))
-            ;
+                RefundTypeInterface::class,
+                implode('", "', $value::keys())
+            ));
         }
 
-        return new RefundType($value);
+        return $this->createType($value);
     }
 
+    /** @param mixed $value */
     public function convertToDatabaseValue($value, AbstractPlatform $platform): ?string
     {
         if (null === $value) {
             return null;
         }
 
         if ($value instanceof RefundType) {
-            return (string) $value;
+            return (string) $value->getValue();
         }
 
-        throw ConversionException::conversionFailed($value, 'sylius_refund_refund_type');
+        throw ConversionException::conversionFailed((string) $value, 'sylius_refund_refund_type');
+    }
+
+    /** @param RefundTypeInterface|string $value */
+    protected function createType($value): RefundTypeInterface
+    {
+        return new RefundType($value);
```

Comment:
This way the implementation is still coupled to `RefundType` instead of `RefundTypeInterface` and its factory.

Reply:
Good idea but this class in not a service, Do you have any idea how to inject factory with param as `RefundType` FQCN?

Reply:
Sadly, we could not be able to do that :( If we're forced to implement it this way, I would add information on how to extend this type in the application that customizes `RefundType` class 🖖

Reply:
Is it mentioned anywhere? If not, I would keep it open :)

Reply:
Its mentioned in PR description and new task for creating docs is ready